packages/host/tests/helpers/adapter.ts

```diff
@@ -59,14 +63,56 @@ export class TestRealmAdapter implements RealmAdapter {
       if (dir.kind === 'file') {
         throw new Error(`tried to use file as directory`);
       }
-      this.#lastModified.set(this.#paths.fileURL(path).href, now);
+      let url = this.#paths.fileURL(path);
+      this.#lastModified.set(url.href, now);
       dir.contents[last] = { kind: 'file', content };
+      if (typeof content === 'object') {
+        this.#potentialModulesAndInstances.push({ content, url });
+      }
+    }
+  }
+
+  get realmPath() {
+    return this.#paths;
+  }
+
+  get ready() {
+    return this.#ready.promise;
+  }
+
+  // We are eagerly establishing shims and preparing instances to be able to be
+  // serialized as our test realm needs to be able to serve these via the HTTP
+  // API (internally) in order to index itself at boot
+  private async prepareInstances() {
+    if (!this.#loader) {
+      throw new Error('bug: loader needs to be set in test adapter');
+    }
+
+    let cardApi = await this.#loader.import<CardAPI>(
+      `${baseRealm.url}card-api`,
+    );
+    for (let { content, url } of this.#potentialModulesAndInstances) {
+      if (cardApi.isCard(content)) {
+        cardApi.setCardAsSavedForTest(
+          content,
+          `${url.href.replace(/\.json$/, '')}`,
+        );
+        continue;
+      }
+      for (let [name, fn] of Object.entries(content)) {
+        if (typeof fn === 'function' && typeof name === 'string') {
+          this.#loader.shimModule(url.href, content);
+          continue;
+        }
+      }
```

Comment:
Note that I still have challenges around the adapter. I was able to solve this without any changes to the `Loader` or `VirtualNetwork`. Ultimately I think this comes down to ordering. We shim the module not because we can't load it, but because we don't have an identity for it when we need it. The challenge is that in the `openFile()` of this adapter, we serialize the requested card instance in the case it was specified as an instance in the adapter's file system. In order to serialize you need to determine the identity for a running card. The challenge is that the identity is only assigned after you import the module. Previously the module was always being imported before we `openFile()` was ever called. Now the order has changed and `openFile()` is called before the module is imported. As such, we call `Loader.shimModule()` eagerly so that we have an identity for the module available by the time `openFile()` is called.